homeassistant/components/stream/hls.py

```diff
@@ -82,29 +109,151 @@ class HlsPlaylistView(StreamView):
     name = "api:stream:hls:playlist"
     cors_allowed = True
 
+    @classmethod
+    def render_current_segment(
+        cls, segment: Segment, ll_hls: bool, last_stream_id: int
+    ) -> str:
+        """Render the most recent segment of a hls playlist including the hint.
+
+        This method is run on every playlist request and the output is not saved.
+        """
+        # Create a copy of the segment with the parts frozen so we can iterate safely
+        segment_copy = copy(segment)
+        segment_copy.parts_by_http_range = segment_copy.parts_by_http_range.copy()
+        (
```

Comment:
I don't think the part byte index is causing any trouble. I actually previously had parts stored separately as a List, and that list still needed to be copied here, so this copy doesn't have to do with the data structure change to a dict at all.
As for the caching of partial responses, I think that's a big optimization. We are going from O(n) to O(1) on parts. With some of my H265+ feeds with really long segments, there are 30+ parts on each segment, and we are carrying parts for ~2.5 segments, so that's ~75 parts we would have to iterate through every time.
Even on regular feeds, depending on settings, there might be 8-10 parts per segment, so that's 20-25 part iterations we are saving.

Reply:
I agree its O(n) vs O(1) but assumed (or maybe was optimistic?) that was too small of an N to matter. Just to clarify: You're pretty confident this is a measurable or noticeable performance win? I realize maybe i run home assistant on normal PC hardware and our target is a raspberry Pi, but i just assumed any CPU related to transcoding was an order of magnitude more than this iteration.

Assuming this performance win stays: did you have any thoughts about the comments about opportunities around code structure w.r.t. to the playlist cache?

Reply:
I haven't profiled it but it makes more sense to me to avoid wasted rendering, especially since this is running on the main thread and it's probably being run every second or less.
We're not doing any transcoding in stream. Do you mean the transmuxing we are doing in the worker? That's done in a different thread and I think the GIL is released for some of it.

Reply:
RE: opportunities around code structure, the render_segment call would be cleaner if it was moved to the Segment object itself, then the Segment object would have HLS related code.

Reply:
Yes, I mis-spoke, was thinking about transmuxing in the worker assuming that is much more CPU time relative to this.  I was thinking about it in terms of overall CPU usage or time from event in camera to event playback in client.

Reply:
I'm not sure actually - I don't think the transmuxing is actually very CPU intense. Neither is string manipulation...but one is looping in python and the other is mostly being done in C. How much this helps does depend on the actual settings used - if n is small then it's of less benefit. As I mentioned, an interesting case is special camera codecs with long GOP/low keyframe intervals such as H264+/H265+. In those cases n can be quite large. Even with low n, I don't see why we wouldn't want to producing playlist in this incremental manner - it seems much more efficient. I think if the rendering was moved to the `Segment` objects this would be more apparent.
While I didn't profile anything, I do remember seeming to notice the effect of this feature. FYI I'm running HA on a 2 core NAS which should have CPU performance about that of a Rpi4.

Reply:
I refactored the hls rendering a bit by unifying the segment rendering into a `Segment.render_hls` method which returns a string with the part of the playlist that corresponds to the `Segment`. The details regarding caching of intermediate data/formatting of the template are hidden in that method - it just gives back a rendered `Segment` as a string, so the calls to that method are a bit cleaner.
We can move that method back from `Segment` to `HlsPlaylistView` if you think we should keep the hls stuff together.